packages/jupyter-ai/jupyter_ai/personas/persona_manager.py

```diff
@@ -0,0 +1,174 @@
+                )
+                self.log.exception(e)
+                continue
+
+        if len(persona_classes) > 0:
+            elapsed_time_ms = (time_ns() - start_time_ns) // 1000
+            self.log.info(
+                f"SUCCESS: Loaded {len(persona_classes)} AI persona classes from entry points. Time elapsed: {elapsed_time_ms}ms."
+            )
+        else:
+            self.log.error(
+                "ERROR: Jupyter AI has no AI personas available. "
+                + "Please verify your server configuration and open a new issue on our GitHub repo if this warning persists."
+            )
+        PersonaManager._persona_classes = persona_classes
+
+    def _init_personas(self) -> dict[str, BasePersona]:
+        """
+        Initializes the list of persona instances for the YChat instance passed
+        to the constructor.
+        """
+        if hasattr(self, "_personas") or len(PersonaManager._persona_classes) == 0:
+            return
+        persona_classes = PersonaManager._persona_classes
+        assert isinstance(persona_classes, list)
+
+        self.log.info(
+            f"PENDING: Initializing AI personas for chat room '{self.ychat.get_id()}'..."
+        )
+        start_time_ns = time_ns()
+
+        personas: dict[str, BasePersona] = {}
+        for Persona in persona_classes:
+            persona = Persona(
+                ychat=self.ychat,
+                manager=self,
+                config=self.config_manager,
+                log=self.log,
+            )
+            if persona.id in personas:
+                class_name = persona.__class__.__name__
+                self.log.warning(
+                    f"  - WARNING: Skipping persona '{persona.name}' from '{persona.__module__}' because another persona has an identical ID '{persona.id}'. "
+                    + f"Personas must all have unique IDs. Please rename the persona class from '{class_name}' to something unique to dismiss this warning."
+                )
+                continue
+            else:
+                self.log.info(
+                    f"  - Initialized persona '{persona.name}' (ID: '{persona.id}')."
+                )
+                personas[persona.id] = persona
+
+        elapsed_time_ms = (time_ns() - start_time_ns) // 1000
+        self.log.info(
+            f"SUCCESS: Initialized {len(personas)} AI personas for chat room '{self.ychat.get_id()}'. Time elapsed: {elapsed_time_ms}ms."
+        )
+        return personas
+
+    @property
+    def personas(self) -> dict[str, BasePersona]:
+        return self._personas
+
+    def get_mentioned_personas(self, new_message: Message) -> list[BasePersona]:
+        mentioned_ids = set(new_message.mentions)
+        persona_list: list[BasePersona] = []
+
+        for mentioned_id in mentioned_ids:
+            if mentioned_id in self.personas:
+                persona_list.append(self.personas[mentioned_id])
+
+        return persona_list
```

Comment:
question: any risk of duplication here?

otherwise you can dedup by using a set to buffer the `mentioned_id` or with:
```py
return list(set(persona_list))
```

Reply:
No risk of duplication since `mentioned_ids` is itself a set, i.e. each persona ID in that set is unique. We also forbid personas having identical IDs in `_init_personas()`:

```py
if persona.id in personas:
    class_name = persona.__class__.__name__
    self.log.warning(
        f"  - WARNING: Skipping persona '{persona.name}' from '{persona.__module__}' because another persona has an identical ID '{persona.id}'. "
        + f"Personas must all have unique IDs. Please rename the persona class from '{class_name}' to something unique to dismiss this warning."
    )
    continue
```